redash/handlers/query_results.py

```diff
@@ -154,6 +156,23 @@ def options(self, query_id=None, query_result_id=None, filetype='json'):
 
         return make_response("", 200, headers)
 
+    def _fetch_rows(self, query_id):
+        query = models.Query.get_by_id_and_org(query_id, self.current_org)
+        query_result = models.QueryResult.get_by_id_and_org(query.latest_query_data_id, self.current_org)
+
+        require_access(query_result.data_source.groups, self.current_user, view_only)
+
+        return json_loads(query_result.data)["rows"]
+
+    def _convert_queries_to_enums(self, definition):
+        if definition["type"] == "query":
+            definition["type"] = "enum"
+
+            rows = self._fetch_rows(definition.pop("queryId"))
+            definition["enumOptions"] = [row.get('value', row.get(row.keys()[0])) for row in rows]
```

Comment:
That's much less code than the client side version 😮 _But_ it's possible that the first column here and on the client side won't be the same one... For now we can ignore this, but we should not wait too long with implementing #3316.

Also to be on the safe side, I would check that rows is not empty.

Reply:
Keep in mind that this code doesn't care about 'name' columns, which is a presentation concern so it won't be as short.